### DISCUSSION TITLE
External DOE + Optimization:  Optuna does not work

### DISCUSSION THREAD

**Discussion by cuismabc** [optuna/optuna#3996] [Q&A] [answered]
### Expected behavior

I tried to import existing experimental results into Optuna's storage by following the suggestions at the following links. I attached two ways..
https://gist.github.com/toshihikoyanase/99f4119f226da4ad227cd6f381136c72
both ways do not work.

1) when I use enqueue, storage and parallelization together, the external samples are not used correctly.
2) when I use alternative way,  
    2.1) I run one optimization directly, good results are obtained.
    2.2) I exported the initial samples out
    2.3) I imported the initial samples into Optuna by using the recommended way,  the optimization results always are not good.

### Environment

- Optuna version: 2.10.0
- Python version: 3.9.6
- OS: Windows X64
- (Optional) Other libraries and their versions:

### Error messages, stack traces, or logs

```shell
for the attached python code,  param_value_pairs includes the first 10 initial samples for one optimization run whose objective function value is 5e-5 (pretty good).

after I used the 10 initial samples as external DOE, the code does not work always..

you can change n_startup_trials=10 or 0 or 1.
```

### Steps to reproduce

import optuna
import random
#https://gist.github.com/toshihikoyanase/99f4119f226da4ad227cd6f381136c72
def suggest_all(trial):
    x = trial.suggest_uniform('x', -10, 10)
    y = trial.suggest_int('y', -2, 2)
    return x, y

def suggest_existing_value(trial, x, y):
    _x = trial.suggest_uniform('x', x, x)
    _y = trial.suggest_int('y', y, y)
    return _x, _y

# Define a simple 2-dimensional quadratic function whose minimum value is -1 when (x, y) = (0, -1).
def quadratic(x, y):
    return x ** 2 + y ** 2

def objective(trial):
    x, y = suggest_all(trial)
    return quadratic(x, y)

def objective_existing_value1(trial, x, y):
    _x, _y = suggest_existing_value(trial, x, y)
    return quadratic(x, y)

def objective_existing_value2(trial, x, y, obj):
    _x, _y = suggest_existing_value(trial, x, y)
    return obj

db_dir = "C:/Users/cuism/Desktop/ex_DOEs/test.db"
# os.remove(db_dir)
storage = ''.join(['sqlite:///', str(db_dir)])

if __name__ == '__main__':
    # Let us minimize the objective function above.
    # Caveat 1: use RandomSampler to avoid issue #318.
    study = optuna.create_study(sampler=optuna.integration.BoTorchSampler( n_startup_trials=10), storage=storage)

    #1. Run grid search.
    param_value_pairs = [{'x': -0.22492644981214127, 'y': 1},
                          {'x': 5.991795279912903, 'y': -1},
                          {'x': 8.060776786638474, 'y': -1},
                          {'x': -4.88345511892585, 'y': -1},
                          {'x': 6.221828475591668, 'y': 1},
                          {'x': -8.787236580946693, 'y': -1},
                          {'x': 5.429724312942881, 'y': 2},
                          {'x': 6.555750112224704, 'y': -1},
                          {'x': 4.287032519784351, 'y': 0},
                          {'x': -4.35059865673667, 'y': -1}]
                        
    for params in param_value_pairs: # can parallel
        obj=quadratic(params['x'],params['y'])
        params['obj']=obj

    for params in param_value_pairs:
        study.optimize(lambda trial: objective_existing_value2(trial, params['x'], params['y'], params['obj']), n_trials=1)

    # print(study.trials_dataframe())
    # print('Best value: {} (params: {})\n'.format(study.best_value, study.best_params))

    # 2. Continue optimization with Optuna.
     
    #study.sampler = optuna.integration.BoTorchSampler( n_startup_trials=10)
    study.optimize(objective, n_trials=30, n_jobs=1)

    print(study.trials_dataframe())
    print('Best value: {} (params: {})\n'.format(study.best_value, study.best_params))

**Comment by nzw0301** [comment, marked as answer]
I'm afraid the code is not executable by others and not readable easily to understand the question. Could you fix it?

**Reply by cuismabc** [reply]
Let me clarity the problem, and please provide one way. Thanks.

We want to minimize  x ** 2 + y ** 2, x is one floating number, and is in the range [-10,10] and y is one integer number, and its range is[-2,2].

One set of external initials are the following and their objective function values have been evaluated. how to pass 10 sets of  x,y,objective function value into Optuna and continue optimization by using TPE/Botorch? 

param_value_pairs = [{'x': -0.22492644981214127, 'y': 1},
                      {'x': 5.991795279912903, 'y': -1},
                      {'x': 8.060776786638474, 'y': -1},
                      {'x': -4.88345511892585, 'y': -1},
                      {'x': 6.221828475591668, 'y': 1},
                      {'x': -8.787236580946693, 'y': -1},
                      {'x': 5.429724312942881, 'y': 2},
                      {'x': 6.555750112224704, 'y': -1},
                      {'x': 4.287032519784351, 'y': 0},
                      {'x': -4.35059865673667, 'y': -1}]

**Reply by nzw0301** [reply]
https://optuna.readthedocs.io/en/stable/reference/generated/optuna.study.Study.html#optuna.study.Study.add_trial

**Reply by cuismabc** [reply]
great.. my problem is solved. thanks
